pkg/dns/aws/dns.go

```diff
@@ -208,48 +195,41 @@ func NewProvider(config Config, operatorReleaseVersion string) (*Provider, error
 		config:    config,
 		idsToTags: map[string]map[string]string{},
 		lbZones:   map[string]string{},
-	}, nil
+	}
+	if err := validateServiceEndpoints(p); err != nil {
+		return nil, fmt.Errorf("failed to validate aws provider service endpoints: %v", err)
+	}
+	return p, nil
 }
 
-// urlContainsValidRegion checks whether uri is valid based on provided region.
-func urlContainsValidRegion(region, uri string) bool {
-	switch {
-	case strings.Contains(uri, Route53Service):
-		switch region {
-		case endpoints.CnNorth1RegionID, endpoints.CnNorthwest1RegionID:
-			if uri == chinaRoute53Endpoint {
-				return true
-			}
-		case endpoints.UsGovEast1RegionID, endpoints.UsGovWest1RegionID:
-			if uri == govCloudRoute53Endpoint {
-				return true
-			}
-		default:
-			if uri == standardRoute53Endpoint || strings.Contains(uri, endpoints.UsEast1RegionID) {
-				return true
-			}
+// validateServiceEndpoints validates that provider clients can communicate
+// with associated API endpoints by each making a list/describe/get call.
+func validateServiceEndpoints(provider *Provider) error {
+	if provider.route53 != nil {
+		input := route53.ListHostedZonesInput{MaxItems: aws.String("1")}
+		if _, err := provider.route53.ListHostedZones(&input); err != nil {
+			return fmt.Errorf("failed to list route53 hosted zones: %v", err)
 		}
-	case strings.Contains(uri, TaggingService):
-		switch region {
-		case endpoints.CnNorth1RegionID, endpoints.CnNorthwest1RegionID:
-			if strings.Contains(uri, endpoints.CnNorthwest1RegionID) {
-				return true
-			}
-		case endpoints.UsGovEast1RegionID, endpoints.UsGovWest1RegionID:
-			if uri == govCloudTaggingEndpoint {
-				return true
-			}
-		default:
-			if strings.Contains(uri, endpoints.UsEast1RegionID) {
-				return true
-			}
+	}
+	if provider.elb != nil {
+		input := elb.DescribeLoadBalancersInput{PageSize: aws.Int64(int64(1))}
+		if _, err := provider.elb.DescribeLoadBalancers(&input); err != nil {
+			return fmt.Errorf("failed to describe elb load balancers: %v", err)
 		}
-	default:
-		if strings.Contains(uri, region) {
-			return true
+	}
+	if provider.elbv2 != nil {
+		input := elbv2.DescribeLoadBalancersInput{PageSize: aws.Int64(int64(1))}
+		if _, err := provider.elbv2.DescribeLoadBalancers(&input); err != nil {
+			return fmt.Errorf("failed to describe elbv2 load balancers: %v", err)
 		}
 	}
-	return false
+	if provider.tags != nil {
+		input := resourcegroupstaggingapi.GetResourcesInput{TagsPerPage: aws.Int64(int64(100))}
+		if _, err := provider.tags.GetResources(&input); err != nil {
+			return fmt.Errorf("failed to get group tagging resources: %v", err)
+		}
+	}
+	return nil
```

Comment:
What do you think of validating all endpoints, accumulating errors, and returning the result at the end?  Something like this:
```go
 	var errs []error
	if _, err := ...; err != nil {
		errs = append(errs, fmt.Errorf(...))
	}
	if _, err := ...; err != nil {
		errs = append(errs, fmt.Errorf(...))
	}
	if _, err := ...; err != nil {
		errs = append(errs, fmt.Errorf(...))
	}
	if _, err := ...; err != nil {
		errs = append(errs, fmt.Errorf(...))
	}
	return kerrors.NewAggregate(errs)
```
(If `errs` is empty, `kerrors.NewAggregate(errs)` returns nil.)

Reply:
Ah, yes I like that approach. Let me update.